typescript/cli/src/config/agent.ts

```diff
@@ -19,19 +23,80 @@ export async function createAgentConfig({
   out,
 }: {
   context: CommandContext;
-  chains: string[];
+  chains?: string[];
   out: string;
 }) {
   logBlue('\nCreating agent config...');
 
-  const { registry, multiProvider, chainMetadata } = context;
+  const { registry, multiProvider, chainMetadata, skipConfirmation } = context;
   const addresses = await registry.getAddresses();
 
-  const core = HyperlaneCore.fromAddressesMap(addresses, multiProvider);
+  await handleNoChainsProvided(skipConfirmation, chains);
 
-  const startBlocks = await promiseObjAll(
-    objMap(addresses, async (chain, _) => {
-      // If the index.from is specified in the chain metadata, use that.
+  const chainAddresses = filterChainAddresses(addresses, chains);
+
+  const core = HyperlaneCore.fromAddressesMap(chainAddresses, multiProvider);
+  const startBlocks = await getStartBlocks(chainAddresses, core, chainMetadata);
+
+  if (!skipConfirmation) {
+    await handleMissingInterchainGasPaymaster(chainAddresses);
+  }
+
+  const agentConfig: AgentConfig = buildAgentConfig(
+    chains ?? Object.keys(chainAddresses),
+    multiProvider,
+    chainAddresses as ChainMap<HyperlaneDeploymentArtifacts>,
+    startBlocks,
+  );
+
+  await validateAgentConfig(agentConfig, skipConfirmation);
+
+  logBlue(`\nWriting agent config to file ${out}`);
+  writeYamlOrJson(out, agentConfig, 'json');
+  logGreen(`✅ Agent config successfully written to ${out}`);
+}
+
+async function handleNoChainsProvided(
+  skipConfirmation: boolean,
+  chains?: string[],
+) {
+  if (!chains || chains.length === 0) {
+    if (skipConfirmation) {
+      logBlue(
+        '\nNo chains provided, generating agent config for all supported chains',
+      );
+    } else {
+      const proceedWithAllChains = await confirm({
+        message:
+          '\nNo chains provided, would you like to generate the agent config for all supported chains?',
+      });
+      if (!proceedWithAllChains) {
+        errorRed('❌ Agent config creation aborted');
+        process.exit(1);
+      }
+    }
+  }
+}
+
+function filterChainAddresses(
+  addresses: ChainMap<ChainAddresses>,
+  chains?: string[],
+) {
+  if (!chains) {
+    return addresses;
+  }
+  return objFilter(addresses, (chain, _): _ is ChainAddresses => {
```

Comment:
we have a nice helper `pick` that does exactly this https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/main/typescript/utils/src/objects.ts#L89